pkg/storage/migrations/migration.go

```diff
@@ -0,0 +1,424 @@
+		if err != nil {
+			return storage.Run{}, err
+		}
+		params = append(params, storage.ValueStrategy(k, stringVal))
+	}
+
+	dest := storage.Run{
+		SchemaVersion:   storage.InstallationSchemaVersion,
+		ID:              src.ID,
+		Created:         src.Created,
+		Namespace:       inst.Namespace,
+		Installation:    src.Installation,
+		Revision:        src.Revision,
+		Action:          src.Action,
+		Bundle:          src.Bundle,
+		BundleReference: src.BundleReference,
+		BundleDigest:    "", // We didn't track digest before v1
+		Parameters:      storage.NewInternalParameterSet(inst.Namespace, src.ID, params...),
+		Custom:          src.Custom,
+	}
+
+	return dest, nil
+}
+
+func (m *Migration) migrateResult(ctx context.Context, inst *storage.Installation, run storage.Run, resultID string) error {
+	// inst is a ref because migrateResult will update the installation status based on the result of the run
+
+	ctx, span := tracing.StartSpan(ctx, attribute.String("resultID", resultID))
+	defer span.EndSpan()
+
+	data, err := m.sourceStore.Read("results", resultID)
+	if err != nil {
+		return span.Error(err)
+	}
+
+	result, err := convertResult(run, data)
+	if err != nil {
+		return span.Error(err)
+	}
+
+	err = m.destStore.Insert(ctx, storage.CollectionResults, storage.InsertOptions{Documents: []interface{}{result}})
+	if err != nil {
+		return span.Error(err)
+	}
+
+	// Update the installation status based on the result of previous runs
+	inst.ApplyResult(run, result)
+
+	// Find all outputs associated with the result
+	outputKeys, err := m.listItems("outputs", resultID)
+	if err != nil {
+		return err
+	}
+
+	for _, outputKey := range outputKeys {
+		if err = m.migrateOutput(ctx, run, result, outputKey); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func convertResult(run storage.Run, data []byte) (storage.Result, error) {
+	var src SourceResult
+	if err := json.Unmarshal(data, &src); err != nil {
+		return storage.Result{}, fmt.Errorf("error parsing result record: %w", err)
+	}
+
+	dest := storage.Result{
+		SchemaVersion:  run.SchemaVersion,
+		ID:             src.ID,
+		Created:        src.Created,
+		Namespace:      run.Namespace,
+		Installation:   run.Installation,
+		RunID:          run.ID,
+		Message:        src.Message,
+		Status:         src.Status,
+		OutputMetadata: src.OutputMetadata,
+		Custom:         src.Custom,
+	}
+
+	return dest, nil
+}
+
+func (m *Migration) migrateOutput(ctx context.Context, run storage.Run, result storage.Result, outputKey string) error {
+	ctx, span := tracing.StartSpan(ctx, attribute.String("outputKey", outputKey))
+	defer span.EndSpan()
+
+	data, err := m.sourceStore.Read("outputs", outputKey)
+	if err != nil {
+		return span.Error(err)
+	}
+
+	output, err := convertOutput(result, outputKey, data)
+	if err != nil {
+		return span.Error(err)
+	}
+
+	// Sanitize sensitive outputs
+	bun := cnab.ExtendedBundle{Bundle: run.Bundle}
+	output, err = m.sanitizer.CleanOutput(ctx, output, bun)
```

Comment:
I'd rather save to the secret store first, and possibly end up with a "dangling" secret than the flip side which would mean that we saved a reference to an external secret that we later find out we couldn't persist, leaving out database in a bad state. If we go with the secret protocol saying that Create really means CreateOrUpdate (upsert) then is this (dangling secrets) still a problem?

If we want to change how the secrets protocol / sanitizer works, I think it makes sense to articular the edge cases that we don't feel are well supported now and then change it in a separate PR so that it gets the consideration it requires.

Reply:
I think if the secret plugins is doing upsert, this is probably ok

Reply:
I'm fixing the k8s plugin to upsert 👍